pkg/utils/utils.go

```diff
@@ -138,3 +143,23 @@ func PrettySlice[T any](s []T, maxItems int) string {
 	}
 	return sb.String()
 }
+
+// GetMaxPods resolves what we should set max pods to for a given nodeclass.
+// If not specified, defaults based on network-plugin. 30 for "azure", 110 for "kubenet",
+// or 250 for "none" and network plugin mode overlay.
+func GetMaxPods(nodeClass *v1alpha2.AKSNodeClass, networkPlugin, networkPluginMode string) int32 {
+	if nodeClass.Spec.MaxPods != nil {
+		return lo.FromPtr(nodeClass.Spec.MaxPods)
+	}
+	if networkPlugin == consts.NetworkPluginNone {
+		return consts.DefaultNetPluginNoneMaxPods
+	}
+	if networkPlugin == consts.NetworkPluginAzure && networkPluginMode == consts.NetworkPluginModeOverlay {
+		return consts.DefaultOverlayMaxPods
+	}
+	if networkPlugin == consts.NetworkPluginAzure && networkPluginMode == consts.NetworkPluginModeNone {
+		return consts.DefaultNodeSubnetMaxPods
+	}
+	// if not an overlay or nodesubnet network-plugin-mode, return the default
+	return consts.DefaultKubernetesMaxPods
```

Comment:
Maybe switch would be more readable? (Left checking nodeClass.Spec.MaxPods out on purpose, seems special)

```suggestion
	switch {
	case networkPlugin == consts.NetworkPluginNone:
		return consts.DefaultNetPluginNoneMaxPods
	case networkPlugin == consts.NetworkPluginAzure && networkPluginMode == consts.NetworkPluginModeOverlay:
		return consts.DefaultOverlayMaxPods
	case networkPlugin == consts.NetworkPluginAzure && networkPluginMode == consts.NetworkPluginModeNone:
		return consts.DefaultNodeSubnetMaxPods
	default:
		return consts.DefaultKubernetesMaxPods
	}
```